File: js/packages/teams-ai/src/planners/AssistantsPlanner.ts
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { TurnContext } from 'botbuilder';
import OpenAI, { AzureClientOptions, AzureOpenAI } from 'openai';

import { AI } from '../AI';
import { TurnState } from '../TurnState';

import { Planner, Plan, PredictedDoCommand, PredictedSayCommand } from './Planner';

/**
 * @private
 */
const DEFAULT_POLLING_INTERVAL = 1000;

/**
 * @private
 */
const DEFAULT_ASSISTANTS_STATE_VARIABLE = 'conversation.assistants_state';

/**
 * @private
 */
const SUBMIT_TOOL_OUTPUTS_VARIABLE = 'temp.submitToolOutputs';

/**
 * @private
 */
const SUBMIT_TOOL_OUTPUTS_MAP = 'temp.submitToolMap';

/**
 * Options for configuring the AssistantsPlanner.
 */
export interface AssistantsPlannerOptions {
    /**
     * The OpenAI or Azure OpenAI API key. Required.
     */
    apiKey: string;

    /**
     * The Azure OpenAI resource endpoint.
     * @remarks
     * Required when using Azure OpenAI. Not used for OpenAI.
     */
    endpoint?: string;

    /**
     * The ID of the assistant to use. Required.
     */
    assistant_id: string;

    /**
     * Optional. Polling interval in milliseconds.
     * @remarks
     * Defaults to 1000 (once a second).
     */
    polling_interval?: number;

    /**
     * Optional. The state variable to use for storing the assistants state.
     * @remarks
     * Defaults to 'conversation.assistants_state'.
     */
    assistants_state_variable?: string;

    /**
     * Optional. Version of the API being called. Default is '2024-02-15-preview'.
     */
    api_version?: string;
}

/**
 * A Planner that uses the OpenAI Assistants API to generate plans for the AI system.
 * @template TState Optional. Type of application state.
 * @remarks
 * This planner manages conversations through OpenAI's thread-based system, handling:
 * - Thread creation and management
 * - Message submission
 * - Tool/function calling
 * - Response processing
 */
export class AssistantsPlanner<TState extends TurnState = TurnState> implements Planner<TState> {
    private readonly _options: AssistantsPlannerOptions;
    private _client: OpenAI;

    /**
     * Creates a new `AssistantsPlanner` instance.
     * @param {AssistantsPlannerOptions} options - Options for configuring the AssistantsPlanner.
     */
    public constructor(options: AssistantsPlannerOptions) {
        this._options = {
            polling_interval: DEFAULT_POLLING_INTERVAL,
            assistants_state_variable: DEFAULT_ASSISTANTS_STATE_VARIABLE,
            api_version: '2024-02-15-preview',
            ...options
        };

        this._client = AssistantsPlanner.createClient(options.apiKey, options.endpoint, this._options);
    }

    /**
     * Starts a new task.
     * @remarks
     * This method is called when the AI system is ready to start a new task. It delegates the
     * task handling to the `continueTask` method. The planner should
     * generate a plan that the AI system will execute. Returning an empty plan signals that
     * there is no work to be performed.
     *
     * The planner should take the users input from `state.temp.input`.
     * @param {TurnContext} context - Context for the current turn of conversation.
     * @param {TState} state - Application state for the current turn of conversation.
     * @param {AI<TState>} ai - The AI system that is generating the plan.
     * @returns {Promise<Plan>} The plan that was generated.
     */
    public async beginTask(context: TurnContext, state: TState, ai: AI<TState>): Promise<Plan> {
        const threadId = await this.ensureThreadCreated(state, context.activity.text);
        await this.blockOnInProgressRuns(threadId);
        return await this.submitUserInput(context, state, ai);
    }

    /**
     * Continues the current task.
     * @remarks
     * This method is called when the AI system is ready to continue the current task. It handles:
     * - Creating a new thread if one doesn't exist
     * - Submitting tool outputs if required
     * - Waiting for any in-progress runs to complete
     * - Submitting user input and creating a new run
     * The method generates a plan that the AI system will execute. Returning an empty plan signals
     * that the task is completed and there is no work to be performed.
     *
     * The output from the last plan step that was executed is passed to the planner via `state.temp.input`.
     * @param {TurnContext} context - Context for the current turn of conversation.
     * @param {TState} state - Application state for the current turn of conversation.
     * @param {AI<TState>} ai - The AI system that is generating the plan.
     * @returns {Promise<Plan>} The plan that was generated.
     */
    public async continueTask(context: TurnContext, state: TState, ai: AI<TState>): Promise<Plan> {
        return await this.submitActionResults(context, state, ai);
    }

    /**
     * Creates a new assistant using the OpenAI Assistants API.
     * @param {string} apiKey - OpenAI API key.
     * @param {OpenAI.Beta.AssistantCreateParams} request - Definition of the assistant to create.
     * @param {string} endpoint - Optional. The Azure OpenAI resource endpoint. Required when using Azure OpenAI.
     * @param {Record<string, string | undefined>} azureClientOptions - Optional. The Azure OpenAI client options.
     * @returns {Promise<OpenAI.Beta.Assistant>} The created assistant.
     * @throws {Error} If the assistant creation fails.
     */
    public static async createAssistant(
        apiKey: string,
        request: OpenAI.Beta.AssistantCreateParams,
        endpoint?: string,
        azureClientOptions?: AzureClientOptions
    ): Promise<OpenAI.Beta.Assistant> {
        const client = AssistantsPlanner.createClient(apiKey, endpoint, azureClientOptions);
        return await client.beta.assistants.create(request);
    }

    /**
     * Ensures a thread exists for the current conversation.
     * @private
     * @param {TurnState} state - The application Turn State.
     * @param {string} input - The initial thread input.
     * @returns {Promise<string>} The thread ID.
     */
    private async ensureThreadCreated(state: TState, input: string): Promise<string> {
        const assistantsState = this.ensureAssistantsState(state);
        if (assistantsState.thread_id == null) {
            const thread = await this._client.beta.threads.create();
            assistantsState.thread_id = thread.id;
            this.updateAssistantsState(state, assistantsState);
        }

        return assistantsState.thread_id;
    }

    /**
     * Checks if a run has reached a terminal state.
     * @private
     * @param {OpenAI.Beta.Threads.Run} run - The run to check.
     * @returns {boolean} True if the run is in a completed state.
     */
    private isRunCompleted(run: OpenAI.Beta.Threads.Run): boolean {
        return ['completed', 'failed', 'cancelled', 'expired'].includes(run.status);
    }

    /**
     * @private
     * Waits for a run to complete.
     * @param {string} thread_id - The ID of the thread.
     * @param {string} run_id - The ID of the run.
     * @param {boolean} handleActions - Whether to handle actions.
     * @returns {Promise<OpenAI.Beta.Threads.Run>} The completed run.
     */
    private async waitForRun(
        thread_id: string,
        run_id: string,
        handleActions = false
    ): Promise<OpenAI.Beta.Threads.Run> {
        while (true) {
            await new Promise((resolve) => setTimeout(resolve, this._options.polling_interval));

            const run = await this._client.beta.threads.runs.retrieve(thread_id, run_id);
            if ((run.status === 'requires_action' && handleActions) || this.isRunCompleted(run)) {
                return run;
            }
        }
    }

    /**
     * @private
     * Submits action results to the assistant.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     * @param {AI<TState>} ai - The AI instance.
     * @returns {Promise<Plan>} A plan based on the action results.
     */
    private async submitActionResults(context: TurnContext, state: TState, ai: AI<TState>): Promise<Plan> {
        const assistantsState = this.ensureAssistantsState(state);

        const actionOutputs = state.temp.actionOutputs;
        const toolMap = state.getValue(SUBMIT_TOOL_OUTPUTS_MAP) as { [key: string]: string };
        const toolOutputs: OpenAI.Beta.Threads.Runs.RunSubmitToolOutputsParams.ToolOutput[] = [];
        for (const action in actionOutputs) {
            const output = actionOutputs[action];
            const toolCallId = toolMap[action];
            if (toolCallId !== undefined) {
                toolOutputs.push({ tool_call_id: toolCallId, output });
            }
        }

        const run = await this._client.beta.threads.runs.submitToolOutputsAndPoll(
            assistantsState.thread_id!,
            assistantsState.run_id!,
            { tool_outputs: toolOutputs }
        );

        const results = await this.waitForRun(assistantsState.thread_id!, run.id, true);
        switch (results.status) {
            case 'requires_action': {
                state.setValue(SUBMIT_TOOL_OUTPUTS_VARIABLE, true);
                return this.generatePlanFromTools(state, results.required_action!.submit_tool_outputs.tool_calls);
            }
            case 'completed': {
                state.setValue(SUBMIT_TOOL_OUTPUTS_VARIABLE, false);
                return this.generatePlanFromMessages(assistantsState.thread_id!, assistantsState.last_message_id!);
            }
            case 'cancelled': {
                return { type: 'plan', commands: [] };
            }
            case 'expired': {
                const expiredCommand: PredictedDoCommand = {
                    type: 'DO',
                    action: AI.TooManyStepsActionName,
                    parameters: {}
                };
                return {
                    type: 'plan',
                    commands: [expiredCommand]
                };
            }
            default:
                throw new Error(
                    `Run failed ${results.status}. ErrorCode: ${results.last_error?.code}. ErrorMessage: ${results.last_error?.message}`
                );
        }
    }

    /**
     * @private
     * Submits user input to the assistant.
     * @param {TurnContext} context - The turn context.
     * @param {TState} state - The turn state.
     * @param {AI<TState>} ai - The AI instance.
     * @returns {Promise<Plan>} A plan based on the user input.
     */
    private async submitUserInput(context: TurnContext, state: TState, ai: AI<TState>): Promise<Plan> {
        const threadId = await this.ensureThreadCreated(state, context.activity.text);

        const message = await this._client.beta.threads.messages.create(threadId, {
            role: 'user',
            content: state.temp.input
        });

        const run = await this._client.beta.threads.runs.createAndPoll(threadId, {
            assistant_id: this._options.assistant_id
        });

        this.updateAssistantsState(state, { thread_id: threadId, run_id: run.id, last_message_id: message.id });
        const results = await this.waitForRun(threadId, run.id, true);
        switch (results.status) {
            case 'requires_action':
                state.setValue(SUBMIT_TOOL_OUTPUTS_VARIABLE, true);
                return this.generatePlanFromTools(state, results.required_action!.submit_tool_outputs.tool_calls);
            case 'completed': {
                state.setValue(SUBMIT_TOOL_OUTPUTS_VARIABLE, false);
                return this.generatePlanFromMessages(threadId, message.id);
            }
            case 'cancelled': {
                return { type: 'plan', commands: [] };
            }
            case 'expired': {
                const expiredCommand: PredictedDoCommand = {
                    type: 'DO',
                    action: AI.TooManyStepsActionName,
                    parameters: {}
                };
                return {
                    type: 'plan',
                    commands: [expiredCommand]
                };
            }
            default:
                throw new Error(
                    `Run failed ${results.status}. ErrorCode: ${results.last_error?.code}. ErrorMessage: ${results.last_error?.message}`
                );
        }
    }

    /**
     * @private
     * Generates a plan from tool calls.
     * @param {TState} state - The turn state.
     * @param {OpenAI.Beta.Threads.Runs.RequiredActionFunctionToolCall[]} toolCalls - The tool calls to generate the plan from.
     * @returns {Plan} A plan based on the tool calls.
     */
    private generatePlanFromTools(
        state: TState,
        toolCalls: OpenAI.Beta.Threads.Runs.RequiredActionFunctionToolCall[]
    ): Plan {
        const plan: Plan = { type: 'plan', commands: [] };
        const toolMap: { [key: string]: string } = {};
        toolCalls.forEach((toolCall) => {
            toolMap[toolCall.function.name] = toolCall.id;
            const doCommand: PredictedDoCommand = {
                type: 'DO',
                action: toolCall.function.name,
                parameters: JSON.parse(toolCall.function.arguments)
            };
            plan.commands.push(doCommand);
        });
        state.setValue(SUBMIT_TOOL_OUTPUTS_MAP, toolMap);
        return plan;
    }

    /**
     * @private
     * Generates a plan from messages.
     * @param {string} thread_id - The ID of the thread.
     * @param {string} last_message_id - The ID of the last message.
     * @returns {Promise<Plan>} A plan based on the messages.
     */
    private async generatePlanFromMessages(thread_id: string, last_message_id: string): Promise<Plan> {
        const messages = await this._client.beta.threads.messages.list(thread_id);

        const lastMessageIndex = messages.data.findIndex((message) => message.id === last_message_id);
        const newMessages = lastMessageIndex >= 0 ? messages.data.slice(0, lastMessageIndex) : [];

        const plan: Plan = { type: 'plan', commands: [] };
        newMessages.forEach((message) => {
            message.content.forEach((content) => {
                if (content.type === 'text') {
                    const sayCommand: PredictedSayCommand = {
                        type: 'SAY',
                        response: {
                            role: 'assistant',
                            content: content.text.value,
                            context: {
                                intent: '',
                                citations:
                                    content.text.annotations?.map((annotation) => ({
                                        title: '',
                                        url: '',
                                        filepath: '',
                                        content: annotation.text
                                    })) || []
                            }
                        }
                    };
                    plan.commands.push(sayCommand);
                }
            });
        });
        return plan;
    }

    /**
     * @private
     * Ensures that the assistants state exists in the turn state.
     * @param {TState} state - The turn state.
     * @returns {AssistantsState} The assistants state.
     */
    private ensureAssistantsState(state: TState): AssistantsState {
        if (!state.hasValue(this._options.assistants_state_variable!)) {
            state.setValue(this._options.assistants_state_variable!, {
                thread_id: null,
                run_id: null,
                last_message_id: null
            } as AssistantsState);
        }

        return state.getValue(this._options.assistants_state_variable!);
    }

    /**
     * @private
     * Updates the assistants state in the turn state.
     * @param {TState} state - The turn state.
     * @param {AssistantsState} assistantsState - The new assistants state.
     */
    private updateAssistantsState(state: TState, assistantsState: AssistantsState): void {
        state.setValue(this._options.assistants_state_variable!, assistantsState);
    }

    /**
     * @private
     * Blocks until all in-progress runs are completed.
     * @param {string} thread_id - The ID of the thread.
     */
    private async blockOnInProgressRuns(thread_id: string): Promise<void> {
        while (true) {
            const runs = await this._client.beta.threads.runs.list(thread_id, { limit: 1 });

            if (runs.data.length === 0) {
                return; // No runs, so we're done
            }
            const latestRun = runs.data[0];
            if (this.isRunCompleted(latestRun)) {
                return; // Latest run is completed, so we're done
            }
            await this.waitForRun(thread_id, latestRun.id);
        }
    }

    /**
     * @private
     * @param {string} apiKey - The api key
     * @param {string} endpoint  - The Azure OpenAI resource endpoint
     * @param {Record<string, string | undefined>} azureClientOptions - The Azure OpenAI client options.
     * @returns {OpenAI} the client
     */
    private static createClient(apiKey: string, endpoint?: string, azureClientOptions?: AzureClientOptions): OpenAI {
        if (endpoint) {
            // Azure OpenAI
            return new AzureOpenAI({ endpoint, apiKey, ...azureClientOptions });
        } else {
            // OpenAI
            return new OpenAI({ apiKey });
        }
    }
}

/**
 * @private
 */
interface AssistantsState {
    thread_id: string | null;
    run_id: string | null;
    last_message_id: string | null;
}

File: python/packages/ai/teams/ai/planners/assistants_planner.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from typing import Dict, Generic, List, Optional, TypeVar

import openai
from botbuilder.core import TurnContext
from dataclasses_json import DataClassJsonMixin, dataclass_json
from openai.types.beta.assistant import Assistant
from openai.types.beta.assistant_create_params import AssistantCreateParams
from openai.types.beta.threads import Run
from openai.types.beta.threads.run import RequiredAction
from openai.types.beta.threads.run_submit_tool_outputs_params import ToolOutput

from ...app_error import ApplicationError
from ...state import TurnState
from ...user_agent import _UserAgent
from ..actions.action_types import ActionTypes
from .plan import Plan, PredictedDoCommand, PredictedSayCommand
from .planner import Planner

DEFAULT_POLLING_INTERVAL = 1
DEFAULT_ASSISTANTS_STATE_VARIABLE = "conversation.assistants_state"
SUBMIT_TOOL_OUTPUTS_VARIABLE = "temp.submit_tool_outputs"
SUBMIT_TOOL_OUTPUTS_MAP = "temp.submit_tool_map"

StateT = TypeVar("StateT", bound=TurnState)


@dataclass_json
@dataclass
class AssistantsState(DataClassJsonMixin):
    thread_id: Optional[str] = ""
    run_id: Optional[str] = ""
    last_message_id: Optional[str] = ""


@dataclass
class AssistantsPlannerOptions:
    """
    Options for configuring the AssistantsPlanner.
    """

    api_key: str
    "The OpenAI API key."

    assistant_id: str
    "The ID of the assistant to use."

    polling_interval: float = DEFAULT_POLLING_INTERVAL
    "Optional. Polling interval in seconds. Defaults to 1 second"

    assistants_state_variable: str = DEFAULT_ASSISTANTS_STATE_VARIABLE
    "Optional. The state variable to use for storing the assistants state."

    organization: Optional[str] = None
    "Optional organization."

    endpoint: Optional[str] = None
    "Optional endpoint."


class AssistantsPlanner(Generic[StateT], _UserAgent, Planner[StateT]):
    "A planner that uses the OpenAI Assistants API."

    _options: AssistantsPlannerOptions
    _client: openai.AsyncOpenAI

    @property
    def options(self) -> AssistantsPlannerOptions:
        return self._options

    @property
    def client(self) -> openai.AsyncOpenAI:
        return self._client

    def __init__(
        self, options: AssistantsPlannerOptions, client: Optional[openai.AsyncOpenAI] = None
    ) -> None:
        """
        Creates a new `AssistantsPlanner` instance.

        Args:
            options (AssistantsPlannerOptions): Options used to configure the planner.
            client (Optional[openai.AsyncOpenAI]): The OpenAI client.

        """

        self._options = options
        self._client = (
            client
            if client is not None
            else openai.AsyncOpenAI(
                api_key=options.api_key,
                organization=options.organization,
                default_headers={"User-Agent": self.user_agent},
                base_url=options.endpoint,
            )
        )

    async def begin_task(self, context: TurnContext, state: TurnState) -> Plan:
        """
        Starts a new task.
        This method is called when the AI system is ready to start a new task. The planner should
        generate a plan that the AI system will execute. Returning an empty plan signals that
        there is no work to be performed. The planner should take the users input
        from `state.temp.input`.

        Args:
            context (TurnContext): Context for the current turn of conversation.
            state (TurnState): Application state for the current turn of conversation.

        Returns:
            Plan: The plan that was generated.
        """
        return await self.continue_task(context, state)

    async def continue_task(self, context: TurnContext, state: TurnState) -> Plan:
        """
        This method is called when the AI system has finished executing the previous plan and is
        ready to continue the current task. The planner should generate a plan
        that the AI system will execute. Returning an empty plan signals that the task is
        completed and there is no work to be performed.
        The output from the last plan step that was executed is passed to the planner
        via `state.temp.input`.

        Args:
            context (TurnContext): Context for the current turn of conversation.
            state (TurnState): Application state for the current turn of conversation.

        Returns:
            Plan: The plan that was generated.
        """

        # Create a new thread id if we don't have one already
        thread_id = await self._ensure_thread_created(state)

        # Add the users input to the thread or send tool outputs
        if state.get(SUBMIT_TOOL_OUTPUTS_VARIABLE) is True:
            # Send the tool output to the assistant
            return await self._submit_action_results(state)

        # Wait for any current runs to complete since you can'tadd messages
        # or start new runs if there's already one in progress
        await self._block_on_in_progress_runs(thread_id)

        # Submit user input
        return await self._submit_user_input(state)

    @staticmethod
    async def create_assistant(
        api_key: str,
        organization: Optional[str],
        endpoint: Optional[str],
        request: AssistantCreateParams,
    ) -> Assistant:
        """
        Static method for programmatically creating an assistant.

        Args:
            api_key (str): The OpenAI API key.
            organization (Optional[str]): The optional organization.
            endpoint: (Optional[str]): The optional endpoint.
            request: (AssistantCreateParams): The parameters used to create the assistant.

        Returns:
            Assistant: The assistant.
        """
        openai_client = openai.AsyncOpenAI(
            api_key=api_key,
            organization=organization if organization else None,
            base_url=endpoint if endpoint else None,
        )

        return await openai_client.beta.assistants.create(
            model=request.get("model", ""),
            description=request.get("description"),
            file_ids=request.get("file_ids", []),
            instructions=request.get("instructions"),
            metadata=request.get("metadata"),
            name=request.get("name"),
            tools=request.get("tools", []),
        )

    async def _ensure_thread_created(self, state: TurnState) -> str:
        assistants_state = self._ensure_assistants_state(state)
        if not assistants_state.thread_id:
            thread = await self._client.beta.threads.create()
            assistants_state.thread_id = thread.id
            state.set(self._options.assistants_state_variable, assistants_state)

        return assistants_state.thread_id

    def _ensure_assistants_state(self, state: TurnState) -> AssistantsState:
        if not state.has(self._options.assistants_state_variable):
            state.set(self._options.assistants_state_variable, AssistantsState())

        assitants_state = state.get(self._options.assistants_state_variable)
        return AssistantsState.from_dict(assitants_state)

    async def _submit_action_results(self, state: TurnState) -> Plan:
        # Get the current assistant state
        assistants_state = self._ensure_assistants_state(state)

        # Map the action outputs to tool outputs
        action_outputs = state.temp.action_outputs
        tool_map = state.get(SUBMIT_TOOL_OUTPUTS_MAP)
        tool_outputs: List[ToolOutput] = []

        for action in action_outputs:
            output = action_outputs[action]
            if tool_map:
                tool_call_id = tool_map[action]
                if tool_call_id is not None:
                    # Add required output only
                    tool_outputs.append(ToolOutput(tool_call_id=tool_call_id, output=output))

        # Submit the tool outputs
        if assistants_state.thread_id and assistants_state.run_id:
            run = await self._client.beta.threads.runs.submit_tool_outputs(
                run_id=assistants_state.run_id,
                thread_id=assistants_state.thread_id,
                tool_outputs=tool_outputs,
            )

            # Wait for the run to complete
            results = await self._wait_for_run(assistants_state.thread_id, run.id, True)
            if results is not None:
                if results.status == "requires_action" and results.required_action:
                    state.set(SUBMIT_TOOL_OUTPUTS_VARIABLE, True)
                    return self._generate_plan_from_tools(state, results.required_action)
                if results.status == "completed" and assistants_state.last_message_id:
                    state.set(SUBMIT_TOOL_OUTPUTS_VARIABLE, False)
                    return await self._generate_plan_from_messages(
                        assistants_state.thread_id, assistants_state.last_message_id
                    )
                if results.status == "cancelled":
                    return Plan()
                if results.status == "expired":
                    return Plan(commands=[PredictedDoCommand(action=ActionTypes.TOO_MANY_STEPS)])

                if results.last_error:
                    raise ApplicationError(
                        f"Run failed {results.status}. ErrorCode: "
                        + f"{results.last_error.code}. ErrorMessage: {results.last_error.message}"
                    )
        return Plan()

    async def _wait_for_run(
        self, thread_id: str, run_id: str, handle_actions: bool = False
    ) -> Optional[Run]:
        while True:
            await asyncio.sleep(self.options.polling_interval)
            run = await self._client.beta.threads.runs.retrieve(thread_id=thread_id, run_id=run_id)
            if run.status == "requires_action":
                if handle_actions:
                    return run
                break
            if run.status == "cancelled":
                return run
            if run.status == "failed":
                return run
            if run.status == "completed":
                return run
            if run.status == "expired":
                return run
        return None

    def _generate_plan_from_tools(self, state: TurnState, required_action: RequiredAction) -> Plan:
        plan = Plan()
        tool_map: Dict = {}
        for tool_call in required_action.submit_tool_outputs.tool_calls:
            tool_map[tool_call.function.name] = tool_call.id
            plan.commands.append(
                PredictedDoCommand(
                    action=tool_call.function.name,
                    parameters=json.loads(tool_call.function.arguments),
                )
            )
        state.set(SUBMIT_TOOL_OUTPUTS_MAP, tool_map)
        return plan

    async def _generate_plan_from_messages(self, thread_id: str, last_message_id: str) -> Plan:
        # Find the new messages
        messages = await self._client.beta.threads.messages.list(thread_id)
        new_messages = []
        for message in messages.data:
            if message.id == last_message_id:
                break
            new_messages.append(message)

        # listMessages return messages in desc, reverse to be in asc order
        new_messages.reverse()

        # Convert the messages to SAY commands
        plan = Plan()
        for message in new_messages:
            for content in message.content:
                if content.type == "text":
                    plan.commands.append(PredictedSayCommand(response=content.text.value))

        return plan

    async def _block_on_in_progress_runs(self, thread_id: str) -> None:
        # We loop until we're told the last run is completed
        while True:
            run = await self._retrieve_last_run(thread_id)
            if not run:
                return
            if self._is_run_completed(run):
                return

            # Wait for the current run to complete and then
            # loop to see if there's already a new run.
            await self._wait_for_run(thread_id, run.id)

    async def _retrieve_last_run(self, thread_id: str) -> Optional[Run]:
        list = await self._client.beta.threads.runs.list(thread_id, limit=1)
        if len(list.data) > 0:
            return list.data[0]
        return None

    def _is_run_completed(self, run: Run) -> bool:
        if run.status == "completed":
            return True
        if run.status == "failed":
            return True
        if run.status == "cancelled":
            return True
        if run.status == "expired":
            return True
        return False

    async def _submit_user_input(self, state: TurnState) -> Plan:
        # Get the current thread_id
        thread_id = await self._ensure_thread_created(state)

        # Add the users input to the thread
        message = await self._client.beta.threads.messages.create(
            thread_id, role="user", content=state.temp.input
        )

        # Create a new run
        run = await self._client.beta.threads.runs.create(
            thread_id=thread_id, assistant_id=self.options.assistant_id
        )

        # Update state and wait for the run to complete
        state.set(
            self._options.assistants_state_variable, AssistantsState(thread_id, run.id, message.id)
        )
        results = await self._wait_for_run(thread_id, run.id, True)

        if results:
            if results.status == "requires_action" and results.required_action is not None:
                state.set(SUBMIT_TOOL_OUTPUTS_VARIABLE, True)
                return self._generate_plan_from_tools(state, results.required_action)
            if results.status == "completed":
                state.set(SUBMIT_TOOL_OUTPUTS_VARIABLE, False)
                return await self._generate_plan_from_messages(thread_id, message.id)
            if results.status == "cancelled":
                return Plan()
            if results.status == "expired":
                return Plan(commands=[PredictedDoCommand(action=ActionTypes.TOO_MANY_STEPS)])

            if results.last_error:
                raise ApplicationError(
                    f"Run failed {results.status}. ErrorCode: "
                    + f"{results.last_error.code}. ErrorMessage: {results.last_error.message}"
                )

        return Plan()
